### ISSUE TITLE
Async request fails when cookies are enabled

### ISSUE THREAD

**Issue by jlricon** [sagebind/isahc#70]
While developing my crate [rlinks](https://github.com/jlricon/rlinks) I found that trying to do async requests fails if cookies are enabled and more than one domain is being requested from. (Maybe there is some collision in the CookieJar; I may look into it)

## Reproducible example

Tested on Ubuntu 19.04, on rustc 1.39.0-nightly (4295eea90 2019-08-30)

with Cargo.toml:
```toml
isahc="0.7"
tokio="0.2.0-alpha.4"
futures-preview="0.3.0-alpha.18"
```
This does work:
```rust
use futures::{stream, StreamExt};
use isahc::{Error, HttpClient};

fn get_client() -> HttpClient {
    HttpClient::builder().cookies().build().unwrap()
}
async fn make_req(client: &HttpClient, url: &str) -> Result<(), Error> {
    client.get_async(url).await;
    Ok(())
}

#[tokio::main]
async fn main() {
    let client = get_client();
    let vec_reqs = vec!["https://google.com", "https://google.com/bla"];
    let res: Vec<Result<(), Error>> = stream::iter(vec_reqs)
        .map(|r| make_req(&client, r))
        .buffer_unordered(10)
        .collect()
        .await;
}
```
But if instead we do
```rust
let vec_reqs = vec!["https://google.com", "https://github.com"];
```
We get instead
```
thread 'main' panicked at 'called `Option::unwrap()` on a `None` value', src/libcore/option.rs:378:21
stack backtrace:
   0: backtrace::backtrace::libunwind::trace
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.35/src/backtrace/libunwind.rs:88
   1: backtrace::backtrace::trace_unsynchronized
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.35/src/backtrace/mod.rs:66
   2: std::sys_common::backtrace::_print
             at src/libstd/sys_common/backtrace.rs:47
   3: std::sys_common::backtrace::print
             at src/libstd/sys_common/backtrace.rs:36
   4: std::panicking::default_hook::{{closure}}
             at src/libstd/panicking.rs:200
   5: std::panicking::default_hook
             at src/libstd/panicking.rs:214
   6: std::panicking::rust_panic_with_hook
             at src/libstd/panicking.rs:477
   7: std::panicking::continue_panic_fmt
             at src/libstd/panicking.rs:384
   8: rust_begin_unwind
             at src/libstd/panicking.rs:311
   9: core::panicking::panic_fmt
             at src/libcore/panicking.rs:85
  10: core::panicking::panic
             at src/libcore/panicking.rs:49
  11: core::option::Option<T>::unwrap
             at /rustc/4295eea903a9e1014ee30f82930f5ec08d888077/src/libcore/macros.rs:12
  12: <isahc::cookies::CookieJar as isahc::middleware::Middleware>::filter_response::{{closure}}
             at /home/jose/.cargo/registry/src/github.com-1ecc6299db9ec823/isahc-0.7.1/src/cookies/mod.rs:294
  13: core::ops::function::impls::<impl core::ops::function::FnMut<A> for &mut F>::call_mut
             at /rustc/4295eea903a9e1014ee30f82930f5ec08d888077/src/libcore/ops/function.rs:265
  14: core::iter::traits::iterator::Iterator::find_map::check::{{closure}}
             at /rustc/4295eea903a9e1014ee30f82930f5ec08d888077/src/libcore/iter/traits/iterator.rs:2008
  15: core::iter::traits::iterator::Iterator::try_for_each::call::{{closure}}
             at /rustc/4295eea903a9e1014ee30f82930f5ec08d888077/src/libcore/iter/traits/iterator.rs:1733
  16: core::iter::adapters::filter_map_try_fold::{{closure}}
             at /rustc/4295eea903a9e1014ee30f82930f5ec08d888077/src/libcore/iter/adapters/mod.rs:918
  17: core::iter::traits::iterator::Iterator::try_fold
             at /rustc/4295eea903a9e1014ee30f82930f5ec08d888077/src/libcore/iter/traits/iterator.rs:1694
  18: <core::iter::adapters::FilterMap<I,F> as core::iter::traits::iterator::Iterator>::try_fold
             at /rustc/4295eea903a9e1014ee30f82930f5ec08d888077/src/libcore/iter/adapters/mod.rs:944
  19: core::iter::traits::iterator::Iterator::try_for_each
             at /rustc/4295eea903a9e1014ee30f82930f5ec08d888077/src/libcore/iter/traits/iterator.rs:1736
  20: core::iter::traits::iterator::Iterator::find_map
             at /rustc/4295eea903a9e1014ee30f82930f5ec08d888077/src/libcore/iter/traits/iterator.rs:2014
  21: <core::iter::adapters::FilterMap<I,F> as core::iter::traits::iterator::Iterator>::next
             at /rustc/4295eea903a9e1014ee30f82930f5ec08d888077/src/libcore/iter/adapters/mod.rs:931
  22: isahc::cookies::CookieJar::add
             at /home/jose/.cargo/registry/src/github.com-1ecc6299db9ec823/isahc-0.7.1/src/cookies/mod.rs:241
  23: <isahc::cookies::CookieJar as isahc::middleware::Middleware>::filter_response
             at /home/jose/.cargo/registry/src/github.com-1ecc6299db9ec823/isahc-0.7.1/src/cookies/mod.rs:300
  24: isahc::client::ResponseFuture::complete::{{closure}}
             at /home/jose/.cargo/registry/src/github.com-1ecc6299db9ec823/isahc-0.7.1/src/client.rs:913
  25: core::result::Result<T,E>::map
             at /rustc/4295eea903a9e1014ee30f82930f5ec08d888077/src/libcore/result.rs:512
  26: isahc::client::ResponseFuture::complete
             at /home/jose/.cargo/registry/src/github.com-1ecc6299db9ec823/isahc-0.7.1/src/client.rs:894
  27: <isahc::client::ResponseFuture as core::future::future::Future>::poll::{{closure}}
             at /home/jose/.cargo/registry/src/github.com-1ecc6299db9ec823/isahc-0.7.1/src/client.rs:940
  28: core::task::poll::Poll<T>::map
             at /rustc/4295eea903a9e1014ee30f82930f5ec08d888077/src/libcore/task/poll.rs:35
  29: <isahc::client::ResponseFuture as core::future::future::Future>::poll
             at /home/jose/.cargo/registry/src/github.com-1ecc6299db9ec823/isahc-0.7.1/src/client.rs:940
  30: std::future::poll_with_tls_context::{{closure}}
             at /rustc/4295eea903a9e1014ee30f82930f5ec08d888077/src/libstd/future.rs:120
  31: std::future::get_task_context
             at /rustc/4295eea903a9e1014ee30f82930f5ec08d888077/src/libstd/future.rs:110
  32: std::future::poll_with_tls_context
             at /rustc/4295eea903a9e1014ee30f82930f5ec08d888077/src/libstd/future.rs:120
  33: check_req::make_req::{{closure}}
             at examples/check_req.rs:8
  34: <std::future::GenFuture<T> as core::future::future::Future>::poll::{{closure}}
             at /rustc/4295eea903a9e1014ee30f82930f5ec08d888077/src/libstd/future.rs:42
  35: std::future::set_task_context
             at /rustc/4295eea903a9e1014ee30f82930f5ec08d888077/src/libstd/future.rs:78
  36: <std::future::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/4295eea903a9e1014ee30f82930f5ec08d888077/src/libstd/future.rs:42
  37: <futures_util::stream::futures_unordered::FuturesUnordered<Fut> as futures_core::stream::Stream>::poll_next
             at /home/jose/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-preview-0.3.0-alpha.18/src/stream/futures_unordered/mod.rs:415
  38: futures_util::stream::StreamExt::poll_next_unpin
             at /home/jose/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-preview-0.3.0-alpha.18/src/stream/mod.rs:1169
  39: <futures_util::stream::buffer_unordered::BufferUnordered<St> as futures_core::stream::Stream>::poll_next
             at /home/jose/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-preview-0.3.0-alpha.18/src/stream/buffer_unordered.rs:118
  40: <futures_util::stream::collect::Collect<St,C> as core::future::future::Future>::poll
             at /home/jose/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-preview-0.3.0-alpha.18/src/stream/collect.rs:51
  41: std::future::poll_with_tls_context::{{closure}}
             at /rustc/4295eea903a9e1014ee30f82930f5ec08d888077/src/libstd/future.rs:120
  42: std::future::get_task_context
             at /rustc/4295eea903a9e1014ee30f82930f5ec08d888077/src/libstd/future.rs:110
  43: std::future::poll_with_tls_context
             at /rustc/4295eea903a9e1014ee30f82930f5ec08d888077/src/libstd/future.rs:120
  44: check_req::main::{{closure}}
             at examples/check_req.rs:16
  45: <std::future::GenFuture<T> as core::future::future::Future>::poll::{{closure}}
             at /rustc/4295eea903a9e1014ee30f82930f5ec08d888077/src/libstd/future.rs:42
  46: std::future::set_task_context
             at /rustc/4295eea903a9e1014ee30f82930f5ec08d888077/src/libstd/future.rs:78
  47: <std::future::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/4295eea903a9e1014ee30f82930f5ec08d888077/src/libstd/future.rs:42
  48: tokio_executor::enter::Enter::block_on
             at /home/jose/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-executor-0.2.0-alpha.4/src/enter.rs:118
  49: tokio::runtime::threadpool::Runtime::block_on::{{closure}}::{{closure}}
             at /home/jose/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.0-alpha.4/src/runtime/threadpool/mod.rs:179
  50: tracing_core::dispatcher::with_default
             at /home/jose/.cargo/registry/src/github.com-1ecc6299db9ec823/tracing-core-0.1.5/src/dispatcher.rs:223
  51: tokio::runtime::threadpool::Runtime::block_on::{{closure}}
             at /home/jose/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.0-alpha.4/src/runtime/threadpool/mod.rs:178
  52: tokio_executor::global::with_default::{{closure}}
             at /home/jose/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-executor-0.2.0-alpha.4/src/global.rs:173
  53: std::thread::local::LocalKey<T>::try_with
             at /rustc/4295eea903a9e1014ee30f82930f5ec08d888077/src/libstd/thread/local.rs:262
  54: std::thread::local::LocalKey<T>::with
             at /rustc/4295eea903a9e1014ee30f82930f5ec08d888077/src/libstd/thread/local.rs:239
  55: tokio_executor::global::with_default
             at /home/jose/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-executor-0.2.0-alpha.4/src/global.rs:147
  56: tokio::runtime::threadpool::Runtime::block_on
             at /home/jose/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.0-alpha.4/src/runtime/threadpool/mod.rs:175
  57: check_req::main
             at examples/check_req.rs:12
  58: std::rt::lang_start::{{closure}}
             at /rustc/4295eea903a9e1014ee30f82930f5ec08d888077/src/libstd/rt.rs:64
  59: std::rt::lang_start_internal::{{closure}}
             at src/libstd/rt.rs:49
  60: std::panicking::try::do_call
             at src/libstd/panicking.rs:296
  61: __rust_maybe_catch_panic
             at src/libpanic_unwind/lib.rs:80
  62: std::panicking::try
             at src/libstd/panicking.rs:275
  63: std::panic::catch_unwind
             at src/libstd/panic.rs:394
  64: std::rt::lang_start_internal
             at src/libstd/rt.rs:48
  65: std::rt::lang_start
             at /rustc/4295eea903a9e1014ee30f82930f5ec08d888077/src/libstd/rt.rs:64
  66: main
  67: __libc_start_main
  68: _start
```

**Comment by sagebind** [comment]
Thanks for identifying this bug! I'll look into this as soon as I can.

**Comment by sagebind** [comment]
Digging deeper, this appears to be a symptom of a larger problem. The cookie middleware needs to examine the effective URI of the response and was coded to do this, but this information isn't actually available. It's safe to say that the cookie jar is effectively broken until this information is made available.

The effective URI is listed as something to add in #47, but actually this information can be added much more easily than the other things listed. I'll work on adding this feature (99% sure this will be called `ResponseExt::effective_uri()`), and then I'll fix this bug here by updating the cookie jar to use this method to get the information it needs.

**Comment by sagebind** [comment]
A fix for this has been published in version 0.7.3!
